Feature/72

subroooo opened the pull request:
## Motivation
- 과제 평가 CRUD api 생성입니다!

<br>

## Key Changes
- lesson-evaluation.(controller, service, dto, swagger...)

- ApiSuccessResponse

- swagger 데코레이터를 위한 .swagger.ts 파일

- LessonEntity 가상 컬럼을 담당하는 프로퍼티 제거
<br>

## To Reviewers
- 석호님께서 원시 타입 및 여러 타입에 대응할 수 있도록 ApiSuccessResponse를 수정해 주셨었는데 이번 pr에 사용할 일이 
있어서 수정하신 코드를 받아 사용했습니다.

-  저번 회의에서 controller layer의 swagger 관련 코드들 때문에 가독성 떨어진다는 이야기가 나와서 각 controller의 
   endpoint마다 .swagger.ts 파일을 만들어 가져다 사용하는 방식으로 변경해 보았습니다.

- LessonEntity에 가상 컬럼을 담당하고 있던 프로퍼티들을 제거해 주었습니다.
  그 이유는 db에 있는 정보들을 뽑아올 때, 가상 컬럼을 join해서 뽑아오는 경우가 적고, 가상 컬럼을 엔티티에 추가해 줄수록 
  Mapped type을 사용하는데 있어 불편함을 겪을 것이기에 엔티티에서 가상 컬럼이 빠지는 것이 낫다고 판단하였습니다.
 
- 이후 작업으로는 과제의 목록 조회 api를 생성하겠습니다

subroooo committed:
:sparkles: #66 - lesson model categoryId

lesson model에 categoryId 추가 및 그에 따른 코드 수정

subroooo committed:
:sparkles: #66 - readSimilarLesson 쿼리 생성

유사 과제 조회 쿼리 생성

subroooo committed:
:sparkles: #66 - readSimilarLesson plainToInstance

유사과제 클라이언트에 넘기기전 캡슐화

subroooo committed:
:white_check_mark: #66 - readSimilarLesson controller

유사 과제 조회 controller layer test code

subroooo committed:
:white_check_mark: #66 - readSimilarLesson service

유사과제 조회 service test code

subroooo committed:
:recycle: #66 - ResponseType > Dto로 네이밍 변경

plainToInstance하는 response  ResponseType > Dto로 네이밍 변경

subroooo committed:
:bug: #66 - develop과 충돌해결

tsconfig 수정된 develop와 충돌해결

subroooo committed:
:recycle: #66 - Lesson Entity tags > hashtag

Lesson엔티티 tags 프로퍼티 hashtag로 수정

subroooo committed:
:bug: #66 - develop과 충돌 해결

develop과 충돌해결

subroooo committed:
:recycle: #66 - 유사 과제 조회 Query String 사용하는 방식으로 변경

Query String Dto를 통한 확장성 있는 유사 과제 조회 api로 변경

subroooo committed:
:fire: #66 - LesssonRepository 상단 주석 삭제

LessonRepository 상단 주석 삭제

subroooo committed:
:sparkles: #70 - 과제 생성,수정,삭제 최소기능 구현

과제 CUD 최소 기능 구현

subroooo committed:
:sparkles: #70 - PrismaHeler, NotFoundErrorFilter

과제의 주인이 맞는지 확인해주기 위한 PrismaHelper,NotFoundErrorFilter 생성

subroooo committed:
:sparkles: #70 - 과제 생성, 수정, 삭제 api 생성 완료

과제 CUD api 생성 완료

subroooo committed:
:white_check_mark: #70 - deleteLesson controller layer

과제 삭제 controller layer test code 추가

subroooo committed:
:white_check_mark: #70 - deleteLesson service layer

과제 삭제 service layer test code

subroooo committed:
:sparkles: #70 - hashtag controller 생성

lesson의 hashtag controller 생성

subroooo committed:
:sparkles: #70 - hashtag 모듈 생성 및 IsRecordInterceptor 수정

라우팅 모듈을 사용해 hashtag 모듈 동적 생성할 수 있게 코드 추가

subroooo committed:
:sparkles: #70 - hashtag entity 생성

hashtag entity 생성

subroooo committed:
:sparkles: #70 - HashtagController,Guard 생성

hashtag controller, IsHostGuard, IsRecordGuard

subroooo committed:
:sparkles: #70 - LessonModule에서 RouterModule사용해 hashtag service 동적 할당

다이나믹 모듈을 활용해 엔드포인트에 따른 다른 service할당

subroooo committed:
:bug: #70 - develop과 충돌 해결

develop과 충돌 해결

subroooo committed:
:recycle: #70 - updateLesson Service layer 오타 수정

updateLesson Service layer test code 오타 수정

subroooo committed:
:sparkles: #70 - lessonhashtag 생성 기능 구현 완료

과제 해시태그 생성 기능 구현 완료

subroooo committed:
:white_check_mark: #70 - createLessonHashtag controller test code

과제 해시태그 컨트롤러 test 코드 작성

subroooo committed:
:white_check_mark: #70 - 과제 createHashtag service test code

과제 해시태그 생성 test code 작성

subroooo committed:
:recycle: #70 - 과제 createHashtag return type 수정

과제 해시태그 생성 return type 수정

subroooo committed:
:recycle: #70 - 호스트가 맞는지 확인하는 로직 service > controller로 위임

호스트가 맞는지 확인하는 로직 service > controller로 위임

subroooo committed:
:white_check_mark: #70 - updateManyHashtag service test code

과제 해시태그 한번에 수정 service layer test code 작성

subroooo committed:
:sparkles: #70 - lesson deleteHashtag, deleteManyHashtag

과제 해시태그 단일삭제 및 대량삭제 코드 구현

subroooo committed:
:sparkles: #70 - 과제 해시태그 단일 수정, 단일 삭제

과제 해시태그 단일 수정, 단일 삭제 기능 추가

subroooo committed:
:white_check_mark: #70 - 과제 updateHashtag controller test code

과제 해시태그 단일 수정 controller test code 작성

subroooo committed:
:white_check_mark: #70 - lesson updateHashtag service test code

과제 해시태그 단일 수정 service layer test code

subroooo committed:
:white_check_mark: #70 - lesson deleteHashtag controller test code

과제 해시태그 단일 삭제 test code 작성

subroooo committed:
:white_check_mark: #70 - lesson deleteHashtag, deleteManyHashtag service test code

과제 해시태그 단일삭제 및 대량삭제 service layer test code 작성

subroooo committed:
:sparkles: #70 - lesson readHashtag 기능 구현 및 controller test 코드 추가

과제 해시태그 조회 기능 구현 및 controller test code 추가

subroooo committed:
:white_check_mark: #70 - lesson readManyHashtag service test code

과제 해시태그 다중 조회 service test code 작성

subroooo committed:
:sparkles: #70 - lesson readHashtag 기능 구현

과제 hashtag 단일 조회 기능 구현

subroooo committed:
:white_check_mark: #70 - lesson readHashtag controller test code

과제 해시태그 단일 조회 controller test code

subroooo committed:
:sparkles: #70 - CustomApiSuccessResponse 생성 및 CustomApiResponse 네이밍 변경

CustomApiResponse > CustomApiFailureResponse 로 네이밍 변경

subroooo committed:
:white_check_mark: #70 - lesson-hashtag controller output test code

과제 해시태그 output 수정에 따른 test 코드 변경

subroooo committed:
:white_check_mark: #70 - lessonHashtag service output에 따른 test code 수정

과제 해시태그 service layer output에 따른 변경

subroooo committed:
:sparkles: #70 - prismaHelper에 validateOwnerOrFail 메서드 추가

validateOwnerOrFail메서드 추가

subroooo committed:
:recycle: #70 - prismaHelper lessonService > lessonController

prismaHelper validateOwnerOrFail controller layer에서 하도록 수정

subroooo committed:
:recycle: #70 - swagger 403,404 message 수정

403,404 swagger 메시지 수정

subroooo committed:
:recycle: #70 - CustomApiResponse 데코레이터 네이밍 변경

CustomApiResponse > Api(Success,Failure)Response

subroooo committed:
:sparkles: #70 - ApiSuccessResponse 생성 및 lesson,lessonHashtag swagger 생    type: LessonHashtagEntity,

lesson, lessonHashtag swagger 작업

subroooo committed:
:recycle: #70 - ApiSuccessResponse 오타 수정

filed > field

subroooo committed:
:recycle: #70 - ApiSuccessResponse 데코레이터 코드 개선

controller 부분 깔끔하게 사용될 수 있도록 코드 개선

subroooo committed:
:recycle: #70 - ApiSuccessResponse 변경에 따른 controller 수정

controller 수정

subroooo committed:
:recycle: #70 - LessonController return 형식 수정

LessonController return 형식 수정

subroooo committed:
:recycle: #70 - validateOwnerOrFail then resolve,reject 에서 reject > catch로 수정

validateOwnerOrFail 에러 핸들링 수정

subroooo committed:
:sparkles: #72 - lesson createEvaluation 기능 구현

남들이 느낀 과제 난이도 평가 기능 구현

subroooo committed:
:sparkles: #72 - lesson createEvaluation swagger 관련 작업

과제 평가 스웨거 관련 작업 완료

subroooo committed:
:white_check_mark: #72 - lesson createEvaluation controller test code

과제 평가 생성 controller layer test code 작성

subroooo committed:
:white_check_mark: #72 - lesson createEvaluation service test code

과제 평가 service layer test code

subroooo committed:
:sparkles: #72 - lesson updateEvaluation 기능 구현

과제 평가 수정 기능 구현

subroooo committed:
:white_check_mark: #72 - lesson updateEvaluation controller test code

과제 평가 수정 controller layer test code

subroooo committed:
:white_check_mark: #72 - lesson updateEvaluation service test code

과제 평가 수정 service layer test code

subroooo committed:
:sparkles: #72 - lesson readEvaluation 기능 구현

과제 평가 조회 기능 구현

subroooo committed:
:recycle: #72 - OptionalJwtAuthGuard 사용하는 부분 type 변경

{ id : null } 추가

subroooo committed:
:recycle: #72 - lesson 모듈 dto 폴더구조 변경

dtos/evaluation, dtos/lesson 이런식으로 생성될 예정

subroooo committed:
:white_check_mark: #72 - lesson readEvaluation controller test code

과제 평가 조회 controller layer test code

subroooo committed:
:white_check_mark: #72 - lesson readMemberEvaluation service test code

멤버가 작성한 과제의 평가 service layer test code

subroooo committed:
:recycle: #70 - seed.ts skipDuplicates

seed 파일에 중복 무시 속성 추가

subroooo committed:
:recycle: #70 - NotFoundException 과 NotFoundError 같은 필터에서 관리되로록 수정

추후 NotFoundFilter 삭제해야 함

subroooo committed:
:recycle: #70 - LessonHashtagParamDto, IdRequestParamDto IsInt 추가

params id 에 float 자료형이 들어와도 대응될 수 있도록 수정

subroooo committed:
:recycle: #70 - HashtagEntity > LessonHashtagEntity

엔티티 클래스명 및 implements추가

subroooo committed:
:recycle: #70 - PrismaHelper delete > PrismaService move method

prismaHelper클래스 제거

subroooo committed:
:fire: #70 - 사용하지 않는 HashtagDto delete

사용하지 않는 dto 삭제

subroooo committed:
:sparkles: #70 - HTTP_ERROR_MESSAGE 추가 및 controller 수정

에러메시지에 대한 swagger 문서 수정

subroooo committed:
:recycle: #70 - lesson-hashtag.controller spec.ts, .ts 구문간 문단 나누기

구문간 문단 나누기

subroooo committed:
:recycle: #70 - lesson.controller 구문간 문단 나누기

구문간 문단 나누기

subroooo committed:
:recycle: #70 - lesson-hashtag.service 구문간 문단 나누기

구문간 문단 나누기

subroooo committed:
:white_check_mark: #70 - lesson.controller test code 수정

과제 controller layer test code 수정

subroooo committed:
:white_check_mark: #70 - lesson.controller test code 수정

test code check method called, check Input & Output 방식으로 수정

subroooo committed:
:white_check_mark: #70 - lesson.service layer test code 수정

check method called, check Input & Output 형식으로 수정

subroooo committed:
:recycle: #70 - lesson-hashtag 메서드명 수정

readHashtag > readOneHashtag, readManyHashtag 형식으로 변경

subroooo committed:
:recycle: #70 - readSimilarLesson 의미없는 변수 할당 제거

의미없는 변수할당 제거

subroooo committed:
:recycle: #70 - dataStructureHelper 메서드 인라인 > 변수로 할당

메서드 중첩 사용 > 변수로 할당해서 사용하는 방식으로 수정

subroooo committed:
:recycle: #70 - lesson-hashtag.controller Promise.all로 비동기 처리

순서 상관없는 비동기 함수 promise.all로 처리

subroooo committed:
:bug: #72 - feature/70 브랜치와 충돌 해결

충돌 해결

subroooo committed:
:recycle: #72 - lesson-evaluation.controller layer return 메시지 수정

컨트롤러 단 swagger 관련 문서 수정

subroooo committed:
:recycle: #70 - PrismaModel의 WhereInput 타입들을 모은 PrismaWhereInput 타입 생성

prismaService 메서드 where 인자 타입 any > PrismaWhereInput으로 변경

subroooo committed:
:memo: #70 - finIneOrFail, PrismaWhereInput 상단에 jsdoc todo주석 추가

jsdoc todo 주석 추가

subroooo committed:
Merge branch 'feature/70' of https://github.com/the-pool/the-pool-api into feature/72

subroooo committed:
:recycle: #72 - lesson module 폴더 구조 변경

과제 모듈 폴더 구조 변경

subroooo committed:
:sparkles: #72 - lesson readManyEvaluation

과제 평가 목록 조회 최소 기능 구현

subroooo committed:
:recycle: #72 - readCountedEvaluation 함수명, 변수명 변경

함수명, 변수명 변경

subroooo committed:
:memo: #72 - readManyEvaluation swagger 관련 코드 추가

LessonEntity에 evaluations 또한 추가 하였음

subroooo committed:
:white_check_mark: #72 - lesson readManyEvaluation controller test code

과제 평가 대량 조회 controller layer test code

subroooo committed:
:white_check_mark: #72 - lesson-evaluation readManyEvaluation service layer test code

과제 평가 대량 조회 service layer test code

subroooo committed:
:sparkles: #72 - lesson Swagger 파일 생성 및 lessonController 수정

lesson swagger 관련 파일 생성

subroooo committed:
:memo: #72 - lessonEvaluation swagger 문서 정리

과제 평가 swagger 파일 생성 및 swagger를 위한 데코레이터 생성

subroooo committed:
:memo: #72 - lessonHashtag swagger 관련 문서 작업

과제 해시태그 swagger 파일 생성 및 데코레이터 작업 완료

subroooo committed:
:white_check_mark: #72 - lesson-hashtag controller test code 수정

과제 해시태그 controller layer test code 수정

subroooo committed:
:white_check_mark: #72 - lessonService readOneLesson test 수정

과제 상세조회 test 수정

subroooo committed:
:recycle: #72 - readCountedEvaluation promise.all로 변경

과제 평가 순서 상관없는 비동기 처리 부분 promise.all로 변경

subroooo committed:
:bug: #72 - develop 과 충돌 해결

충돌 해결

subroooo committed:
:bug: #72 - develop 과 충돌 해결.

충돌해결

subroooo committed:
:fire: #72 - lessonHashtagParamDto 삭제

필요 없는 param dto 삭제

subroooo committed:
:memo: #72 - lesson swagger applyDecorators

과제 swagger applyDecorators 추가

RokwonK commented on src/decorators/api-success-response.decorator.ts:
```diff
@@ -1,31 +1,158 @@
+
 export const ApiSuccessResponse = (
   status: Exclude<HttpStatus, ErrorHttpStatusCode>,
-  props: { [key: string]: Type },
+  referenceType: Record<string, Reference>,
+  primitiveType?: Record<string, Primitive>,
 ) => {
-  const extraModels: Type[] = [];
-  const properties: Record<string, ReferenceObject> = {};
+  const extraModels: Type<unknown>[] = [];
+  const properties: Record<string, SchemaObject | ReferenceObject> = {};
+  const apiResponseOption: ApiResponseMetadata | ApiResponseSchemaHost = {
+    status,
+    schema: {},
+  };
+
```
이 파일 코드 이해를 못해서 회의때 한 번 설명 부탁드릴게요!

subroooo replied:
@rrgks6221 석호님께서 만드신거라 석호님이 잘 설명해 주실 겁니다!

RokwonK commented on src/modules/lesson/services/lesson-evaluation.service.ts:
```diff
@@ -0,0 +1,100 @@
+
+  /**
+   * lesson의 체감 난이도를 생성한 유저의 체감난이도 조회
+   */
+  readMemberEvaluation(
+    lessonId: number,
+    memberId: number | null,
+  ): Promise<LessonEvaluationEntity | null> | null {
+    if (memberId === null) {
+      return null;
+    }
+
+    return this.prismaService.lessonLevelEvaluation.findFirst({
+      where: { lessonId, memberId },
+    });
+  }
```
Lesson에 담겨있는게 좋지 않을까요?
예를 들어, Lesson을 리스트를 불러왔을때 item마다 체감난이도를 확인하기 위해선 Lesson 자체에 담겨있는게 좋을 것 같은데 어떻게 생각하시나요?

subroooo replied:
Lesson entity에 LessonLevelEvaluation의 컬렉션이 담겨있는게 좋을 것 같다는 의견이신가요
아니면 위 메서드가 LessonService 클래스에 담겨 있는게 좋을 것 같다는 의견이신가요?

RokwonK commented on src/modules/lesson/swaggers/lesson-hashtag.swagger.ts:
```diff
@@ -0,0 +1,71 @@
+import { applyDecorators, HttpStatus } from '@nestjs/common';
+import { ApiOkResponse, ApiOperation } from '@nestjs/swagger';
+import { HTTP_ERROR_MESSAGE } from '@src/constants/constant';
+import { ApiFailureResponse } from '@src/decorators/api-failure-response.decorator';
+import { ApiSuccessResponse } from '@src/decorators/api-success-response.decorator';
+import { LessonHashtagEntity } from '../entities/lesson-hashtag.entity';
+
+export const ApiCreateManyHashtag = (summary: string) => {
+  return applyDecorators(
+    ApiOperation({ summary }),
+    ApiSuccessResponse(HttpStatus.CREATED, {
+      hashtags: { type: LessonHashtagEntity, isArray: true },
+    }),
+    ApiFailureResponse(HttpStatus.FORBIDDEN, HTTP_ERROR_MESSAGE.FORBIDDEN),
+    ApiFailureResponse(HttpStatus.NOT_FOUND, HTTP_ERROR_MESSAGE.NOT_FOUND),
+  );
+};
```
깔끔합니다!
아직까지 단점은 생각나지 않지만 좋은 시도 같습니다!

subroooo replied:
좋게 봐주셔서 감사드립니다!
이번주 백엔드 정기 회의에서 조금 더 이야기 해보죠!

RokwonK reviewed (approved):
PR을 최소기능단위로 쪼갤 수 있도록 회의해보죠!
고생하셨습니다!

rrgks6221 commented on src/modules/lesson/controllers/lesson-evaluation.controller.spec.ts:
```diff
@@ -0,0 +1,211 @@
+    let memberId: number;
+    let updatedEvaluation: LessonEvaluationEntity;
+
+    beforeEach(async () => {
```
```suggestion
    beforeEach( () => {
```

subroooo replied:
beforeEach 이후 무지성 async 습관 잘 안고쳐지네요 ㅠㅠ 신경 쓰겠습니다

rrgks6221 commented on src/modules/lesson/controllers/lesson-evaluation.controller.spec.ts:
```diff
@@ -0,0 +1,211 @@
+    let countedEvaluation: LessonLevelEvaluationEntity[];
+    let memberEvaluate: LessonEvaluationEntity | null;
+
+    beforeEach(async () => {
```
```suggestion
    beforeEach(() => {
```

rrgks6221 commented on src/modules/lesson/controllers/lesson-evaluation.controller.ts:
```diff
@@ -0,0 +1,125 @@
+export class LessonEvaluationController {
+  constructor(
+    private readonly lessonEvaluationService: LessonEvaluationService,
+    private readonly prismaHelper: PrismaService,
```
```suggestion
    private readonly prismaService: PrismaService,
```

subroooo replied:
오... 역시 예리하십니다 감사해요!

rrgks6221 commented on src/modules/lesson/dtos/evaluation/lesson-evaluation-param.dto.ts:
```diff
@@ -0,0 +1,18 @@
+  @IsRecord({ model: ModelName.LessonLevelEvaluation, field: 'id' }, true)
+  @Min(1)
+  @Type(() => Number)
+  evaluationId: number;
```
이 프로퍼티에 float 형 오면 어떻게 되는지 확인해주세요

subroooo replied:
IsInt를 빼먹었네요.. 감사합니다

rrgks6221 commented on src/modules/lesson/dtos/evaluation/lesson-evaluation-query.dto.ts:
```diff
@@ -0,0 +1,28 @@
+  @Min(1)
+  @IsRecord<Member>({ model: 'member', field: 'id' }, true)
+  @Type(() => Number)
+  memberId: number;
```
1. 이 프로퍼티에 float 타입 오면 어떻게 되는지 확인해주세요
2. 옵셔널 프로퍼티에는 옵셔널 타입 추가해주세요
3. 데코레이터 순서 한번 확인해주세요
queryString 쪽으로 넘어오는 값에 대해 DB 에 정보가 없다면 에러를 줄지 빈 배열을 줄지 한번 다같이 얘기해보면 좋을거같아요
@RokwonK

subroooo replied:
말씀하시는 옵셔널 타입이 DTO클래스에 프로퍼티에 물음표가 붙는 선택적 프로퍼티 말씀하시는게 맞을까요?
```
memberId : number > memberId? :  number
```

저는 우선 빈 배열을 return 하도록 설계하였고,
 ```
@IsRecord<Member>({ model: 'member', field: 'id' }, true)
@Min(1)
@IsInt()
@Type(() => Number)
```
순으로 데코레이터 재배치하였습니다!

rrgks6221 replied:
네 Isoptional 데코레이터가 붙는다는게 클라이언트에서 주는 값 중 null 과 undefined 를 허용한다는 의미니까요

rrgks6221 commented on src/modules/lesson/dtos/evaluation/lesson-evaluation-query.dto.ts:
```diff
@@ -0,0 +1,28 @@
+  @IsOptional()
+  @IsEnum(LessonLevelId)
+  @Type(() => Number)
+  levelId: LessonLevelId;
```
옵셔널 타입 지정해주세요

rrgks6221 commented on src/modules/lesson/dtos/lesson/update-evaluation.dto.ts:
```diff
@@ -0,0 +1,14 @@
+    nullable: true,
+  })
+  @IsOptional()
+  levelId: LessonLevelId;
```
옵셔널인지 널러블인지 정확히 표현해주세요

subroooo replied:
nullable은 빼도록 하겠습니다!

rrgks6221 commented on src/modules/lesson/services/lesson-evaluation.service.ts:
```diff
@@ -0,0 +1,100 @@
+  /**
+   * lesson의 체감 난이도 변경
+   */
+  async updateEvaluation(
```
이거 하나의 레슨에 유저는 하나의 평가만 할 수 있디는 가정이면 delete 후에 create 하는거보다 update 하는게 나을거겉은데 지금 로직 채택한 이유가 ㅇㅆ나요??

subroooo replied:
아직 과제 평가 플로우가 정확하게 정해지지 않아서 과제 평가를 수정하는 api를 만드는데 있어 고민을 많이 했습니다.

POST로 되어 있는 과제 평가 생성 api는 과제물 제출 이후에 과제 채감 난이도를 평가할 수 있는 페이지가 생긴다면 해당 페이지에서 과제 평가 생성을 위한 api이고,
PUT으로 되어있는 과제 평가 수정 api는 과제물을 작성 완료한 유저가 해당 과제를 상세조회 한 이후 평가 난이도를 변경하는 경우를 위한 api입니다.

과제물 제출 이후에 과제 평가 여부가 필수가 아닌 경우에 클라이언트 단에서 비효율적인 분기 처리가 들어갈 것으로 예상되어 lessonEvaluation의 id를 몰라도 lesson와 member만으로 요청이 처리될 수 있도록 delete 후 create로 하였습니다.

위에서 말씀드린 비효율적인 분기 처리란
유저가 과제 평가를 한 뒤 과제 상세 조회를 했을 때 평가를 "상" > "하" 로 변경하는 경우 클라이언트 개발자는 기존에 
있는 lessonEvaluation id를 사용해 id에 해당하는 evaluation level을 "하" 로만 update 하면 되지만

**과제를 마친 뒤 과제 평가를 하지 않은 유저가 뒤늦게 과제 상세조회를 한 뒤 과제 평가를 하고 싶은 경우**에는 lessonEvaluation id가 없어 create 해주는 POST 요청을 보내야 하는 분기 처리가 발생합니다. 

위처럼 분기 처리가 발생하는 부분에서 저는 과제물을 작성한 유저가 과제 평가를 1회만 할 수 있기 때문에 과제 고유 id와  유저 고유 번호만을 이용한 put 요청을 사용해 delete 후 create 하는 방식이 더 좋다고 판단하여 delete 후 create 하는 방식으로 진행하였습니다.

rrgks6221 replied:
이 부분은 대면으로 한번 얘기해보도록 하죠

rrgks6221 commented on src/decorators/api-success-response.decorator.ts:
```diff
@@ -1,31 +1,158 @@
+  }
+
+  return mapStr + ' }';
+};
```
얘네들은 데코레이터도 아니고 swagger 를 만들기 위한 util 성 함수들이고 또 다른 곳에서 사용할 수 있으니까 다른 파일에 있고 가져와 사용하는게 좋을거같아요

subroooo replied:
좋습니다 common.ts에 옮겨 놓도록 하겠습니다.

rrgks6221 commented on src/decorators/api-success-response.decorator.ts:
```diff
@@ -1,31 +1,158 @@
+  return mapStr + ' }';
+};
+
+type OptionalProperty = {
+  isArray?: boolean;
+  nullable?: boolean;
+  deprecated?: boolean;
+  title?: string;
+  description?: string;
+  maxLength?: number;
+  minLength?: number;
+  maxItems?: number;
+  minItems?: number;
+  uniqueItems?: boolean;
+  maxProperties?: number;
+  minProperties?: number;
+  enum?: any[];
+  default?: any;
+};
+
+type Reference = OptionalProperty & {
+  type: Type<unknown>;
+};
+
+type Primitive = OptionalProperty & {
+  type: string | Record<string, any>;
+  example?: any;
+};
```
얘네들도 별도의 type 파일에서 관리해서 export 하여 사용하는게 좋을거같아요.

subroooo replied:
type.ts에 옮겨 놓도록 하겠습니다!

rrgks6221 commented on src/modules/core/database/prisma/prisma.service.ts:
```diff
@@ -91,4 +92,20 @@ export class PrismaService
       });
   }
+
+  async validateDuplicateAndFail(
```
이 아이 설명 한번만 해주세요

subroooo replied:
과제 평가 생성처럼 유저가 특정 리소스를 오직 한 개만 생성할 수 있는 경우 리소스를 생성하기 이전에 db에 중복된 데이터가 있는지 확인해주는 메서드입니다.

rrgks6221 replied:
설명 감사합니다.
1. validateDuplicateAndFail 어떤게 중복인지 모호한 메서드명인거같아요
2. model 이 변수명이 맞는지 한번 생각해봐주시면 좋겠어요
3. 설명해주신대로 한개만 생성할 수 있다는걸 validate 하는거라면 findMany 를 사용하는게 맞나요???
4. findMany 를 사용한다면 model?.length 처럼 옵셔널체이닝을 걸 필요가 있나요??
5. throw new ConflictException(`${modelName} is duplicatd`); 여기에 duplicated 오타있습니다.

rrgks6221 replied:
이거 깃헙 이슈로 다른데에 코멘트가 달린거같은데 where 조건문때문에 findMany 를 사용한거라면 부적절하다고 생각해요 만약 findMany 를 한 뒤에 해당 값을 리턴 받아서 사용하면 모르겠는데 validate를 위해 다소 무거운 findMany 보다는 다른 방법을 생각해보는게 좋지 않을까 싶어요
findFirst 나 count 로 조회하는 방법도 고려해주시면 좋을거같아요

rrgks6221 commented on src/modules/lesson/entities/lesson-evaluation.entity.ts:
```diff
@@ -0,0 +1,27 @@
+  })
+  lessonId: number;
+
+  @ApiProperty({
```
보니까 dto 쪽에서 IsEnum 으로 걸려있던데 그러면 타입도 enum 으로 변경되어야할거같고 어떤 방법으로든 클라이언트 측에 enum mapping 정보를 알려줘야할거같네여.

subroooo replied:
맞습니다 제가 꼼꼼하지 못했네요

DTO에 있는 levelId 프로퍼티는 
```
@ApiPropertyOptional({
  example: LessonLevelId.Top,
  description: '과제를 수행한 member가 느끼는 과제의 난이도',
  enum: getEntriesByEnum(LessonLevelId),
})
@IsEnum(LessonLevelId)
@IsOptional()
levelId: LessonLevelId;
```
로 변경하도록 하였고,

lesson-evaluation.entity는 
```
@ApiProperty({
  example: 1,
  description: '과제를 수행한 member가 느끼는 과제의 난이도',
  enum: getEntriesByEnum(LessonLevelId),
})
levelId: LessonLevelId;
```
로 변경하였습니다!

rrgks6221 commented on src/modules/lesson/services/lesson-evaluation.service.ts:
```diff
@@ -0,0 +1,100 @@
+  readManyEvaluation(
+    lessonId: number,
+    { memberId, levelId }: LessonEvaluationQueryDto,
+  ) {
```
타입 지정 빠졌어요.

subroooo replied:
타입 추가 완료하였습니다!

rrgks6221 commented on src/modules/lesson/controllers/lesson-evaluation.controller.ts:
```diff
@@ -0,0 +1,125 @@
+    @SetModelNameToParam(ModelName.Lesson)
+    param: IdRequestParamDto,
+    @Query() query: LessonEvaluationQueryDto,
+  ): Promise<any> {
```
any 타입은 가급적 사용하지 말아주세요

subroooo replied:
LessonEntity에 가상 컬럼을 빼면서 타입이 깨져서 임시로 any로 넣어놨었는데 다시 채우는 것을 빼먹었군요! 꼼꼼하게 봐주셔서 감사해요

rrgks6221 commented on src/modules/lesson/entities/lesson-level-evaluation.entity.ts:
```diff
@@ -1,25 +1,13 @@
   })
-  @Type(() => Number)
-  bottom: number;
+  count: number;
```
LessonLevelEvaluation 에는 count 라는 컬럼이 없는데 이 클레스 명이 Entity 로 지어야하는지 조금 의문이 드네요

rrgks6221 commented on src/modules/lesson/services/lesson-evaluation.service.ts:
```diff
@@ -0,0 +1,100 @@
+    const countedEvaluation: LessonLevelEvaluationEntity[] = [];
+
+    lessonEvaluations.forEach((lessonEvaluation) => {
+      const obj = <LessonLevelEvaluationEntity>{};
+
+      obj['levelId'] = lessonEvaluation.levelId;
+      obj['count'] = lessonEvaluation._count.lessonId;
```
이 세줄 뭔가 사탄들렸다고 생각드는데 기분탓인가요??

subroooo replied:
오우... 저 왜 저렇게 했죠??? 피곤했나?

rrgks6221 commented:
![image](https://user-images.githubusercontent.com/46591459/211791490-58599500-cb58-49a2-af49-5fae5a74b951.png)
요거도 펼쳐서 확인해주세요

subroooo committed:
:recycle: #72 - optional property, class-validator 순서 지정

DTO 클래스 옵셔널 프로퍼티, class-validator 순서 변경

subroooo committed:
:recycle: #72 - api-success-response.decorator에 있는 common 함수 common.ts 로 이동

common 함수 이동

subroooo committed:
:recycle: #72 - api-success-response.decorator.ts type > type.ts로 이동

type 이동

subroooo committed:
:recycle: #72 - update-evaluaion, lesson-evaluation.entity 프로퍼티 수정

프로퍼티 수정

subroooo committed:
:recycle: #72 - lesson-evaluation 타입추가

controller any > {evaluations:LessonEvaluationEntity[]}, service Promise<LessonEvaluationEntity[]>

rrgks6221 commented on src/constants/constant.ts:
```diff
@@ -28,4 +28,5 @@ export const MAJOR_TEXT = {
 export const HTTP_ERROR_MESSAGE = {
   FORBIDDEN: 'You do not have access to {modelName}',
   NOT_FOUND: "{modelId} doesn't exist id in {modelName}",
+  CONFLICT: '{model} is duplicatd',
```
```suggestion
  CONFLICT: '{model} is duplicated',
```

subroooo replied:
찾아주셔서 감사합니다!

rrgks6221 commented on src/modules/lesson/services/lesson-evaluation.service.ts:
```diff
@@ -0,0 +1,100 @@
+    lessonId: number,
+    levelId: number | null,
+    memberId: number,
+  ): Promise<LessonEvaluationEntity | {}> {
```
혹시 vscode 에 eslint 관련해서 warning 띄워주는 설정이 안돼있나요??
에디터에서 {} 보다는 Record<string, never> 를 사용하는게 어떤지 물어보네요
<img width="699" alt="image" src="https://user-images.githubusercontent.com/46591459/212046797-a3e7dc08-d9fb-42d3-a09f-ed94b150c27c.png">

```ts
const obj1: {} = 'asd';
```
같은 코드를 넣었을 때 {} 를 any 로 받아들여서 빌드도 성공되네요
@RokwonK 님도 참고해주시면 좋을거같아요

subroooo replied:
좋은 정보 감사드립니다!
```
{} => Record<string, never>
```
타입 변경하도록 하겠습니다.

오타 수정하였고, 옵셔널 체이닝은 빼주었습니다. (where 조건에 걸리는 row가 없으면 빈 배열을 return 해주더라구요)
추가로 findMany 메서드가 PrismaModel을 return하니 변수명을 prismaModels이 어울릴 것 같아 변경하였습니다.

validate 해주는 메서드를 처음에는 prisma 빌트인 메서드인 findUniqueOrThrow을 사용하려고 했지만 findUniqueOrThrow는 해당 model의 고유 id가 where 절에 필수로 들어가야 하는 메서드라 
**WhereUniqueInput needs at least one argument and at least one argument for id.** 에러가 발생하게 됩니다.

따라서 findMany를 통해서 배열로 뽑은 이후 length 속성을 통해서 validate해주었습니다.

rrgks6221 commented:
> PR을 최소기능단위로 쪼갤 수 있도록 회의해보죠! 고생하셨습니다!

이 부분 미리 죄송합니다...

subroooo committed:
:recycle: #72 - validateDuplicateAndFail 변수명 변경 및 옵셔널 체이닝 제거

변수명 변경, 옵셔널 체이닝 제거

subroooo committed:
:recycle: #72 - readCountedEvaluation 사탄들린 부분 수정

사탄들린 코드 수정

subroooo committed:
:recycle: #72 - @Type, class-transfomer, enum을 활용해 readCountedEvaluation 로직 변경

과제 평가 횟수 조회 로직 변경

subroooo committed:
:memo: #72 - ReadEvaluationDto countedEvaluation type 수정

dto ApiProperty 타입 수정

rrgks6221 commented on src/modules/lesson/dtos/evaluation/update-evaluation.dto.ts:
```diff
@@ -0,0 +1,16 @@
+  })
+  @IsEnum(LessonLevelId)
+  @IsOptional()
+  levelId: LessonLevelId;
```
optional type 빠졌네요 늦게 발견해서 ㅈㅅ

subroooo replied:
일관성있게 levelId 프로퍼티를 받는 것이 좋은 것 같다고 생각되어 Optional에서 nullable로 수정하였고
```
@IsIn([...getValueByEnum(LessonLevelId, 'number'), null])
```
을 통해 enum 또는 null만 허용된 값으로 받을 수 있도록 하였습니다!

rrgks6221 reviewed (commented):
과제 평가쪽만 대면으로 조금 얘기해보죠

subroooo committed:
:recycle: #72 - valicateDuplicateAndFail findMany > findFirst, UpdateEvaluationDto

UpdateEvaluationDto optional > nullable로 변경

subroooo committed:
:memo: #72 - schema.prisma LessonLevelEvaluation 연관관계 추가

levelId 연관관게 추가

subroooo committed:
:recycle: #72 - updateEvaluationDto class-validator 수정

class-validator 수정
